### PR TITLE
[TR][PAM] PIM-4603: Add the asset category tree in the product asset picker

### PR DESCRIPTION
**Pull request by phaseinducer**
https://akeneo.atlassian.net/browse/PIM-4603

| Q                  | A
| ------------------ | ---
| Migration script?  | N/A
| Behats             | no regression
| Specs & Static     | no regression
| Changelog updated  | N/A

### CODE REVIEW

File: features/Context/WebUser.php [hunk 1]
```diff
@@ -1614,6 +1614,24 @@ public function iCheckTheSwitch($status, $locator)
...
+     * @When /^I (en|dis)able the inclusion of sub-categories$/
+     */
+    public function iSwitchTheSubCategoriesInclusion($status)
+    {
+        $switch = $this->spin(function () {
+            return $this->getCurrentPage()->findById('nested_switch_input');
+        }, 5);
+
+        $on = 'en' === $status;
+        if ($on != $switch->isChecked()) {
```
Comment (jmleroux) [inline features/Context/WebUser.php]: `$on !== $switch->isChecked()`

**Comment by grena** [comment]
GTM

File: features/Context/Page/Product/Edit.php [hunk 2]
```diff
@@ -1367,6 +1375,40 @@ public function selectTree($category)
...
     /**
+     * @param Category $category
+     *
+     * @throws \Exception
+     */
+    public function clickCategoryFilterLink($category)
+    {
+        $node = $this
+            ->getCategoryTree()
+            ->find('css', sprintf('#node_%s a', $category->getId()));
```
Comment (jmleroux) [inline features/Context/Page/Product/Edit.php]: no need to spin ?

File: features/Context/Page/Product/Edit.php [hunk 3]
```diff
@@ -1390,12 +1432,12 @@ public function expandCategory($category)
      */
     public function findCategoryInTree($category)
     {
-        $elt = $this->getElement('Category tree')->find('css', sprintf('li a:contains("%s")', $category));
-        if (!$elt) {
+        $leaf = $this->getCategoryTree()->find('css', sprintf('li a:contains("%s")', $category));
+        if (!$leaf) {
```
Comment (jmleroux) [inline features/Context/Page/Product/Edit.php]: `null !== $leaf`

File: features/Context/WebUser.php [hunk 4]
```diff
@@ -1614,6 +1614,24 @@ public function iCheckTheSwitch($status, $locator)
...
+            $switch->getParent()->find('css', 'label')->click();
+        }
+        $this->wait();
```
Comment (jmleroux) [inline features/Context/WebUser.php]: wait for what ?
not possible to make an explicit `wait` ?

**Comment by jjanvier** [comment]
reviewed but not opinion on that, sorry @phaseinducer

File: features/Context/Page/Product/Edit.php (hunk 2 above)
Comment (phaseinducer) [inline features/Context/Page/Product/Edit.php]: No, the tree is already loaded at this point.

File: features/Context/WebUser.php (hunk 4 above)
Comment (phaseinducer) [inline features/Context/WebUser.php]: We need to wait for the tree reloading here, the default conditions are fine for that.

File: features/Context/Page/Product/Edit.php [hunk 5]
```diff
@@ -1369,6 +1377,40 @@ public function selectTree($category)
...
+    /**
+     * Filter by unclassified products
+     */
+    public function clickUnclassifiedCategoryFilterLink()
+    {
+        $node = $this
+            ->getCategoryTree()
+            ->find('css', '#node_-1 a');
+
+        if (!$node) {
```
Comment (solivier) [inline features/Context/Page/Product/Edit.php]: is node a boolean?

File: features/Context/Page/Product/Edit.php [hunk 6]
```diff
@@ -1392,12 +1434,12 @@ public function expandCategory($category)
      */
     public function findCategoryInTree($category)
     {
-        $elt = $this->getElement('Category tree')->find('css', sprintf('li a:contains("%s")', $category));
-        if (!$elt) {
+        $leaf = $this->getCategoryTree()->find('css', sprintf('li a:contains("%s")', $category));
+        if (!$leaf) {
```
Comment (solivier) [inline features/Context/Page/Product/Edit.php]: same

File: features/Context/Page/Product/Edit.php [hunk 7]
```diff
@@ -1587,4 +1629,20 @@ protected function findScopedField($name, $scope)
...
         return $this->find('css', sprintf('#%s', $scopeLabel->getAttribute('for')));
     }
+
+    /**
+     * @return NodeElement|null
+     */
+    public function getCategoryTree()
+    {
+        $modal = $this->find('css', '.modal');
+        if ($modal &&
```
Comment (solivier) [inline features/Context/Page/Product/Edit.php]: fit in one line

**Comment by solivier** [comment]
GTM when fixed

**Pull request merged**